### PR TITLE
Changes for edge-21.1.3

### PR DESCRIPTION
**Pull request by cpretzer**
## edge-21.1.3

This edge release improves proxy diagnostics and handling for the failfast state
by improving log output and enhancing the proxy's ability to recover when it is
in failfast. In addition, the commands and sub-commands for `linkerd viz` have
been updated to be consistent with the `linkerd` core commands. Also, the
install commands for the core and multicluster components have been modified to
accept helm-style customization flags.

Full release notes:

* Added helm `set`, `set-string`, `values`, `set-files` customization flags for
  the `linkerd install` and `linkerd multicluster install` commands
* Fixed `linkerd metrics` resource selector to differentiate between owner
  resource names
* Added tap-injector to linkerd-viz which is responsible for adding the tap
  service name environment variable to the Linkerd proxy container
* Improved proxy diagnostics about the proxy's failfast state
* Made failfast recovery for a service more robust when new requests are being
  received
* Improved metric labeling for TCP server metrics, as well as inbound and
  outbound tls metrics
* Added `client` and `server` prefixes to socket-level errors to indicate which
  side of the proxy encountered the error
* Added a watch to jaeger-injector to prevent errors in environments with high
  pod or namespace churn
* Added check to confirm whether the jaeger injector pod is in running state
  (thanks @yashvardhan-kukreja!)
* Fixed some logic that caused a crash when EndpointSlices were enabled
  (thanks @oleh-ozimok!)
* Added a `linkerd viz check` sub-command to verify the states of the
  `linkerd-viz` components
* Added a `log-format` flag to optionally output the control plane component log
  output as JSON (thanks @mo4islona!)
* Added logic to use the configured default namespace for `metrics` and
  `profile` subcommands

Signed-off-by: Charles Pretzer <charles@buoyant.io>

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,6 +1,43 @@
...
 # Changes
 
+## edge-21.1.3
+
+This edge release improves proxy diagnostics and handling for the failfast state
+by improving log output and enhancing the proxy's ability to recover when it is
+in failfast. In addition, the commands and sub-commands for `linkerd viz` have
+been updated to be consistent with the `linkerd` core commands. Also, the
+install commands for the core and multicluster components have been modified to
+accept helm-style customization flags.
```
Comment (olix0r) [inline CHANGES.md:11]: This is a little unclear to me. Perhaps a higher-level summary like:

```suggestion
in failfast. Additionally, the `viz` and `multicluster` CLI sub-commands have
been updated for consistency. 
```

File: CHANGES.md [hunk 2]
```diff
@@ -1,6 +1,43 @@
...
+
+Full release notes:
+
+* Added helm `set`, `set-string`, `values`, `set-files` customization flags for
+  the `linkerd install` and `linkerd multicluster install` commands
+* Fixed `linkerd metrics` resource selector to differentiate between owner
+  resource names
+* Added tap-injector to linkerd-viz which is responsible for adding the tap
+  service name environment variable to the Linkerd proxy container
+* Improved proxy diagnostics about the proxy's failfast state
```
Comment (olix0r) [inline CHANGES.md:21]:
```suggestion
* Improved diagnostics about the proxy's failfast state
```

File: CHANGES.md [hunk 3]
```diff
@@ -1,6 +1,43 @@
...
+* Made failfast recovery for a service more robust when new requests are being
+  received
+* Improved metric labeling for TCP server metrics, as well as inbound and
+  outbound tls metrics
```
Comment (olix0r) [inline CHANGES.md:25]:
```suggestion
  outbound TLS metrics
```

File: CHANGES.md [hunk 4]
```diff
@@ -1,6 +1,43 @@
 
 # Changes
 
+## edge-21.1.3
+
+This edge release improves proxy diagnostics and handling for the failfast state
```
Comment (adleong) [inline CHANGES.md:6]: Probably most readers are unfamiliar with what failfast means.  What about something like: "... improves proxy diagnostics and recovery in situations where the proxy is temporarily unable to route requests". ? @olix0r wdyt?

File: CHANGES.md [hunk 5]
```diff
@@ -1,6 +1,43 @@
...
+by improving log output and enhancing the proxy's ability to recover when it is
+in failfast. In addition, the commands and sub-commands for `linkerd viz` have
+been updated to be consistent with the `linkerd` core commands. Also, the
+install commands for the core and multicluster components have been modified to
+accept helm-style customization flags.
+
+Full release notes:
+
+* Added helm `set`, `set-string`, `values`, `set-files` customization flags for
```
Comment (adleong) [inline CHANGES.md:15]: s/helm/Helm-style

File: CHANGES.md [hunk 6]
```diff
@@ -1,6 +1,43 @@
...
+* Improved metric labeling for TCP server metrics, as well as inbound and
+  outbound tls metrics
+* Added `client` and `server` prefixes to socket-level errors to indicate which
+  side of the proxy encountered the error
+* Added a watch to jaeger-injector to prevent errors in environments with high
+  pod or namespace churn
+* Added check to confirm whether the jaeger injector pod is in running state
+  (thanks @yashvardhan-kukreja!)
+* Fixed some logic that caused a crash when EndpointSlices were enabled
+  (thanks @oleh-ozimok!)
```
Comment (olix0r) [inline CHANGES.md:33]:
```suggestion
* Fixed a crash in the destination controller when EndpointSlices were enabled
  (thanks @oleh-ozimok!)
```

File: CHANGES.md [hunk 7]
```diff
@@ -1,6 +1,43 @@
...
+in failfast. In addition, the commands and sub-commands for `linkerd viz` have
+been updated to be consistent with the `linkerd` core commands. Also, the
+install commands for the core and multicluster components have been modified to
+accept helm-style customization flags.
+
+Full release notes:
+
+* Added helm `set`, `set-string`, `values`, `set-files` customization flags for
+  the `linkerd install` and `linkerd multicluster install` commands
+* Fixed `linkerd metrics` resource selector to differentiate between owner
```
Comment (adleong) [inline CHANGES.md:17]: Fixed an issue where `linkerd metrics` could return metrics for the incorrect set of pods when there are overlapping label selectors

File: CHANGES.md [hunk 8]
```diff
@@ -1,6 +1,43 @@
...
+  resource names
+* Added tap-injector to linkerd-viz which is responsible for adding the tap
+  service name environment variable to the Linkerd proxy container
+* Improved proxy diagnostics about the proxy's failfast state
+* Made failfast recovery for a service more robust when new requests are being
+  received
+* Improved metric labeling for TCP server metrics, as well as inbound and
```
Comment (adleong) [inline CHANGES.md:24]: Improved how?

File: CHANGES.md [hunk 9]
```diff
@@ -1,6 +1,43 @@
...
+  outbound tls metrics
+* Added `client` and `server` prefixes to socket-level errors to indicate which
```
Comment (adleong) [inline CHANGES.md:26]: I'm not sure what this refers to.  Log messages? Metric labels?

File: CHANGES.md [hunk 10]
```diff
@@ -1,6 +1,43 @@
...
+  side of the proxy encountered the error
+* Added a watch to jaeger-injector to prevent errors in environments with high
```
Comment (adleong) [inline CHANGES.md:28]: Added watch RBAC permissions to the jaeger-injector to prevent...

File: CHANGES.md [hunk 11]
```diff
@@ -1,6 +1,43 @@
...
+* Added check to confirm whether the jaeger injector pod is in running state
+  (thanks @yashvardhan-kukreja!)
+* Fixed some logic that caused a crash when EndpointSlices were enabled
+  (thanks @oleh-ozimok!)
+* Added a `linkerd viz check` sub-command to verify the states of the
+  `linkerd-viz` components
+* Added a `log-format` flag to optionally output the control plane component log
+  output as JSON (thanks @mo4islona!)
+* Added logic to use the configured default namespace for `metrics` and
+  `profile` subcommands
```
Comment (olix0r) [inline CHANGES.md:39]: Not clear to me. Did we fix a problem in the prior edge? If so, perhaps "Fixed an issue where the `metrics` and `profile` subcommands used the incorrect viz namespace"?

File: CHANGES.md [hunk 12]
```diff
@@ -1,6 +1,43 @@
...
+* Improved proxy diagnostics about the proxy's failfast state
+* Made failfast recovery for a service more robust when new requests are being
+  received
+* Improved metric labeling for TCP server metrics, as well as inbound and
+  outbound tls metrics
+* Added `client` and `server` prefixes to socket-level errors to indicate which
+  side of the proxy encountered the error
+* Added a watch to jaeger-injector to prevent errors in environments with high
+  pod or namespace churn
+* Added check to confirm whether the jaeger injector pod is in running state
```
Comment (adleong) [inline CHANGES.md:30]: jaeger-injector

File: CHANGES.md [hunk 13]
```diff
@@ -1,6 +1,43 @@
...
+  service name environment variable to the Linkerd proxy container
+* Improved proxy diagnostics about the proxy's failfast state
+* Made failfast recovery for a service more robust when new requests are being
+  received
+* Improved metric labeling for TCP server metrics, as well as inbound and
+  outbound tls metrics
+* Added `client` and `server` prefixes to socket-level errors to indicate which
+  side of the proxy encountered the error
+* Added a watch to jaeger-injector to prevent errors in environments with high
+  pod or namespace churn
```
Comment (olix0r) [inline CHANGES.md:29]: Perhaps
```suggestion
* Improved jaeger-injector reliability in environments with many resources
```

File: CHANGES.md (hunk 4 above)
Comment (olix0r) [reply CHANGES.md:6]: +1

File: CHANGES.md (hunk 8 above)
Comment (olix0r) [reply CHANGES.md:24]: We can probably omit this. It's not user-facing in any of our tools yet.

**Review by adleong — APPROVED** [review]
💯

**Review by olix0r — APPROVED** [review]

**Pull request merged**